### ISSUE TITLE
Remote clients connections fail in GKE

### ISSUE THREAD

**Issue by jsanda** [k8ssandra/k8ssandra-operator#82]
We use kubeconfig files to create client connections. For a kind cluster that looks something like this:

```yaml
apiVersion: v1
clusters:
  - cluster:
      certificate-authority-data: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUM1ekNDQWMrZ0F3SUJBZ0lCQURBTkJna3Foa2lHOXcwQkFRc0ZBREFWTVJNd0VRWURWUVFERXdwcmRXSmwKY201bGRHVnpNQjRYRFRJeE1EZ3hNakUxTWpVd01sb1hEVE14TURneE1ERTFNalV3TWxvd0ZURVRNQkVHQTFVRQpBeE1LYTNWaVpYSnVaWFJsY3pDQ0FTSXdEUVlKS29aSWh2Y05BUUVCQlFBRGdnRVBBRENDQVFvQ2dnRUJBUEt1CkFTVUdCOVhJUlZnNkFUN0QvczcwVU9JQUdGNEFnMmFsc1hlSjFRL2lDMEM5c1hlM3V4TGlzV0RCZHNyZzMrNkkKQUFldVo2NmxpTnduREJtNHRnRDlLZ2lMeVBnN2RuT2sxNEZHNHgzRmlxRUtRV1dkSW5sc0gxT2NIdWdqM1R4WgpJeUxLUlMxWWJNc2kybkVPcVdJNGxxbTNmcjgvdFRBUHZqeWhFT2VVYVNnMGhCZTJRbDloeGFTQU9sSXp3NjJEClU2WHFwcGdvY2QyaithNVBiZWo0VTFlZk9icGFjUTRUMElyd2lYNzN4ZnNPendldG9yQnB2OFBRczBKVGMrWngKYTlyZTc1TmxTRVdXRlVVNXVkK2xYMXA2Y3pzc3BwRzA3dkJjakQ0NHBvenpiTXM5c2pNdWdVaGRCUWRMbjYweQpZRnhXYlZPSHR0WVgzb1FEd3hjQ0F3RUFBYU5DTUVBd0RnWURWUjBQQVFIL0JBUURBZ0trTUE4R0ExVWRFd0VCCi93UUZNQU1CQWY4d0hRWURWUjBPQkJZRUZHZXQ4RFZFTDhwdmtncmIvV2VIdmpRbkJBcjRNQTBHQ1NxR1NJYjMKRFFFQkN3VUFBNElCQVFBcFpKS3pscmpYa01TYzVmeGZqcG5NUC83cStGSzcxUUw4cm9RZ0ZuLy9DNlUyLzRGWQpnck1nbUdUTCtMdnVUdXJJTjh0cDhGQ2tDYWxSVVNuMTRtUHFsTHdXby93ZEpNVUFTeCsrdzFtbGkzRldDQ2FLCmlyR1NpU3kwLzEwMDExR3Y3elhRZ3dvYmltSUxXSnJjWldGNFNXRUx2bGZvRUhLYXd0NlIzWVJjbUY5NjJzS2sKYlRpaW1CU2ZUQ2RHa0lKYmxhb2x6WWp5Yi9hbHdMZVIxbHRLZXl6dURjNXdaeU1EOWk0SDRXYlhZaDUya3NnOQpxYnRDcjhWallZL0owYm5HdGxMb0Q5KzEvTkxnanNaNW1pRkhWbnR5N2hYRHJLYTJEcmxQNmQycVp3RXNlWlZiCkQ3bFpTdm9HZkcxb25yN0Q0bURaVHZ6MTJxTXhsbHhQQlp0ZAotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg==
      server: https://172.18.0.6:6443
    name: kind-k8ssandra-0
contexts:
  - context:
      cluster: kind-k8ssandra-0
      user: kind-k8ssandra-0
    name: kind-k8ssandra-0
current-context: kind-k8ssandra-0
kind: Config
preferences: {}
users:
  - name: kind-k8ssandra-0
    user:
      client-certificate-data: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURJVENDQWdtZ0F3SUJBZ0lJTkZtb29DTGU1WUl3RFFZSktvWklodmNOQVFFTEJRQXdGVEVUTUJFR0ExVUUKQXhNS2EzVmlaWEp1WlhSbGN6QWVGdzB5TVRBNE1USXhOVEkxTURKYUZ3MHlNakE0TVRJeE5USTFNRFJhTURReApGekFWQmdOVkJBb1REbk41YzNSbGJUcHRZWE4wWlhKek1Sa3dGd1lEVlFRREV4QnJkV0psY201bGRHVnpMV0ZrCmJXbHVNSUlCSWpBTkJna3Foa2lHOXcwQkFRRUZBQU9DQVE4QU1JSUJDZ0tDQVFFQXFwbm01bDE2THpjejdldkgKNlA4MHNVZ2Z6UmdtZWZQYnozVmMzZ0FzdTltQ2sxbitBZkdzQUl2WGFkUC9YUmExalVQcUFtY0R3Q0RsTkt5VAo3aC9qK1VkbGhMWERHRmxlVjRURld3a0RpNy9SWmJoSXBuVE1iUHFYRlg2eFJLbjVlRWtVTWRjWGRpbytGWVloClhmQzdyajdmbm40RXZnU3VmR1NTand2QndFUmppVEJWdkRuME9KOFFRbXdpVEh0VFhNWGQva1JRSExqUkw2L1kKZVFuL05vYytaVzBncTd3Zm9MTzlOSmozV0pxbTR4UmtXNkxGV29PMXhVRWJCcVh1K0JNYldIbU0xKytXVUpkTAp6UTlhSXVTcXQrOFVIKzlscDU4SXZZYk42c3RPdVVSeGtzaHBmVVpQcWhDNzhxU2RUQ0V2b0plUzVDUUlBbUxYCmdnQTdEUUlEQVFBQm8xWXdWREFPQmdOVkhROEJBZjhFQkFNQ0JhQXdFd1lEVlIwbEJBd3dDZ1lJS3dZQkJRVUgKQXdJd0RBWURWUjBUQVFIL0JBSXdBREFmQmdOVkhTTUVHREFXZ0JSbnJmQTFSQy9LYjVJSzIvMW5oNzQwSndRSworREFOQmdrcWhraUc5dzBCQVFzRkFBT0NBUUVBY2MzUndmSzBHSnBoS3NWT3ZvZzU5TExzcWVvaFYvYlZpclJ6CkQzdHM2U3NUVmo1ZUZpQ3hZUndXNWF2c3ROZ3ltcmF6Njg3UDhJMThxVDZyUGhNanZrUm9sMEVjVEp6cDJxUWoKclA2ZzNzWW1CcmtRVFRIc3VPcW9xckJDSzFIRVhnMm5qc1ZrWnllQncrSXNQM1RlWlVvQWcxZGlSd0M2TWI3VQpCa3ZhQmFPUDNxQVNRMHhSVzFMNStPRzhBemZjc1Q2bHR3Z3RrUWdXSHFBV09Yd0d5VVUzTGZxSzBsZlJtODJNClV0UGIwdTE0SWw3dVhIRmRMdTIwc3Z0VEhxNDZmWE02aXJzQTFVNlhzSll3NUhtTmZLUUNVV0sxUThzNGptUHUKb3dIc2lVNkVlTVVBR0gvK244Y0grUmF0d0U2YXJLemt4QWU2YWtqSmdwRTJGc3Q1aGc9PQotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg==
      client-key-data: LS0tLS1CRUdJTiBSU0EgUFJJVkFURSBLRVktLS0tLQpNSUlFcFFJQkFBS0NBUUVBcXBubTVsMTZMemN6N2V2SDZQODBzVWdmelJnbWVmUGJ6M1ZjM2dBc3U5bUNrMW4rCkFmR3NBSXZYYWRQL1hSYTFqVVBxQW1jRHdDRGxOS3lUN2gvaitVZGxoTFhER0ZsZVY0VEZXd2tEaTcvUlpiaEkKcG5UTWJQcVhGWDZ4UktuNWVFa1VNZGNYZGlvK0ZZWWhYZkM3cmo3Zm5uNEV2Z1N1ZkdTU2p3dkJ3RVJqaVRCVgp2RG4wT0o4UVFtd2lUSHRUWE1YZC9rUlFITGpSTDYvWWVRbi9Ob2MrWlcwZ3E3d2ZvTE85TkpqM1dKcW00eFJrClc2TEZXb08xeFVFYkJxWHUrQk1iV0htTTErK1dVSmRMelE5YUl1U3F0KzhVSCs5bHA1OEl2WWJONnN0T3VVUngKa3NocGZVWlBxaEM3OHFTZFRDRXZvSmVTNUNRSUFtTFhnZ0E3RFFJREFRQUJBb0lCQVFDSEJZbU8xM3JsdFo0ZQo5dFpnbk9ONlVDZ0RMZEwrQURmL1NOYjRTVkVlVjdjVm9rcjhCT2JxQThPNDJYV3BtbGlVZHFUVncreGUrbW9OCnQyelpNVnBBSi9hOEF4Tm15UXgyM2ZsSTVLaHV5YUVVRGJkS0x2VEFaOUI5emFqRlBIWGFjcFJPSWZ0S3BZSS8Kak1vSnpJYnVFSXJ4cUxGZUp0eG1UVDhGZ2Z1VmkvL2duS2crUVUrb0tFWkFieThWLzYwUEZrQ3BlODJIVnZoYQpKM01qZGhRZENwWFdod1ZwRm1vWjNJMVhDT3JubTg5UzNTSjM1cWp3RlNHbTdONWlIQkVRRmMvbE9veWJZU2xWCmpuS0FBYzV4My80eVpCVFZYRkVidUM3cllMYnNQVzBHOWZwK0NOdEpkVUZSUWFxeTBHYlF3eW5wR3lJL0xCTnkKYktjNFBvRGhBb0dCQU5sWXpwTTFZY1JocTNPT3dPRTJpNTJ0YnZ3Tmk2czNoQlF5bTI0VTVlek50c0RocHozRQp1T0hWdDJTQUdEMzRqeFBwV1VPS0RWQzlqK09qU2dSNWsxVTREak5kWVUrbEhWbStLWW9BU0s5RlhsWERQMW50Ck1hd3MzdEFzOE02ZzEyZjhzODhKQ2k0Y2tBazZ1UEZFcVBRVjRYdnJ2eU9tNjZ0WGVSWXpIeWI1QW9HQkFNancKNU0ybTNRT2xFRld6ZGFNZzh0dTRHUzljSzROYzNyUTFMRHprSWVjQ3lmMm1ZR1hTT3VjejlMU1JIU3VVampabApsWnliSmtFYmZSdTV6YmpZZHgzSXV5QVZKR0dYNUVySVgveVc1MGR3NkNOMTZwM2l6TWFJbzFEQmdpMytiRU1sCnJUQ0YwOWZNSURQQnBjaHFzbDJVVDFXWVE0RDJCMTAyaDBFQ2tWVzFBb0dBSjlmZDFxS2NqNUdRQ3BrS3hyNW4KRGJxemM4VlA2MDA4Q3kvVHluUk9TZ0RVRmJ2SUdzeXJUOHNMMllLWjd4Slh1RUM0dW1iZVZnVEtFZUxuT0xiSQp5WUNES2V0ZVlXbFZYaFE4TFM0c0pPVkVpOWhGRHNrcXAvTjV6UlBQK1RFSEMrc0pVMVRHNXZrWmRuZXJXNTlJCmthWkpOWnhiNTc4N3lLYzZiVlZseXBFQ2dZRUFvMm1NbVZaWExPeWk4UjlGenNrdjE5SEs0WUlTZlAwY3UrNzIKTC9yVE8xeEFTb0J3cXIyaVczNGNSbkF4Q2MwQkxtbkVseHh4cXNjaElsZ2pYbXpqeWh3QldlRDN1a3BkUlJmWgpQOEMxQzV0ZG5lRDZVVFl0bEZOd2hXOHBZQ0FsaVNIdHk2VTZGNmdRbENhcW5PZHdjTHY1SE1IOE83ZFpXOHZKCmlPditwM1VDZ1lFQXVwZnhuQXZYOWJzSnRHYy9kVktvTU9lUUk4NE5BaXN6UG40LzZoNVJlS2Q1ekNPNVRWU1kKbXgvcUZ6amxrZStDYmM5OFJad1dWZzJTS3FtazhJbDJYU1kyOUpEWUZnUHJBWWVteGZmVnR4N1R2K3Jua25GZApNU0Y4K0NCRHg2QzM5aTh1UC9HSzUxaUtOSTRjaDhudlN5QUI2N3lHVGl5cHY3djBZNFVQeWhvPQotLS0tLUVORCBSU0EgUFJJVkFURSBLRVktLS0tLQo=
```

For clusters running in managed clouds, there will be some cloud-specific auth. Here is a kubeconfig for a GKE cluster:

```yaml
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    server: https://1.2.3.4
  name: my_gke_cluster
contexts:
- context:
    cluster: my_gke_cluster
    user: my_gke_cluster
  name: my_gke_cluster
current-context: my_gke_cluster
kind: Config
preferences: {}
users:
- name: my_gke_cluster
  user:
    auth-provider:
      config:
        access-token: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        cmd-args: config config-helper --format=json
        cmd-path: /path/to/gcloud/google-cloud-sdk/bin/gcloud
        expiry: "2021-08-12T21:54:29Z"
        expiry-key: '{.credential.token_expiry}'
        token-key: '{.credential.access_token}'
      name: gcp
```

In GKE that auth token expires every hour. I would assume that the expiration is configurable but I have not checked. The operator is able to successfully create a client connection with this kubeconfig. After an hour though, we start seeing errors in the log like this:

```
E0813 01:21:47.301404       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.20.2/tools/cache/reflector.go:167: Failed to watch *v1beta1.CassandraDatacenter: failed to list *v1beta1.CassandraDatacenter: Get "https://35.237.6.160/apis/cassandra.datastax.com/v1beta1/namespaces/default/cassandradatacenters?resourceVersion=109631": error executing access token command "/Users/johnsanda/gcloud/google-cloud-sdk/bin/gcloud config config-helper --format=json": err=fork/exec /path/to/gcloud/google-cloud-sdk/bin/gcloud: no such file or directory output= stderr=
E0813 01:22:03.163765       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.20.2/tools/cache/reflector.go:167: Failed to watch *v1.Pod: failed to list *v1.Pod: Get "https://35.237.6.160/api/v1/namespaces/default/pods?resourceVersion=110817": error executing access token command "/path/to/gcloud/google-cloud-sdk/bin/gcloud config config-helper --format=json": err=fork/exec /path/to/gcloud/google-cloud-sdk/bin/gcloud: no such file or directory output= stderr=
```

This is not an issue when running locally outside of kubernetes with kubectl for example assuming the `gcloud` tool is available.

https://www.fullstory.com/blog/connect-to-google-kubernetes-with-gcp-credentials-and-pure-golang describes an approach for authenticating that does not require using a kubeconfig. An approach like this requires cloud-specific API calls. 

I think that we can solve this in a cloud-agnostic way by using service account tokens. Every pod is associated with a service account. If a service account is not explicitly specified for a pod then namespace default will be used. Pods authenticate against the api server using the service account token which is automatically mounted in the pod. You can read more about this [here](https://kubernetes.io/docs/reference/access-authn-authz/authentication/).

When creating an in-cluster connection, client-go expects to find the service account token at `/var/run/secrets/kubernetes.io/serviceaccount/token` and the CA cert at `/var/run/secrets/kubernetes.io/serviceaccount/ca.crt`. The api server url is provided via the `KUBERNETES_SERVICE_HOST`, `KUBERNETES_SERVICE_PORT` env vars. With this client-go can create a `rest.Config` object.

Here is the relevant function from client-go:

```go
func InClusterConfig() (*Config, error) {
	const (
		tokenFile  = "/var/run/secrets/kubernetes.io/serviceaccount/token"
		rootCAFile = "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
	)
	host, port := os.Getenv("KUBERNETES_SERVICE_HOST"), os.Getenv("KUBERNETES_SERVICE_PORT")
	if len(host) == 0 || len(port) == 0 {
		return nil, ErrNotInCluster
	}

	token, err := ioutil.ReadFile(tokenFile)
	if err != nil {
		return nil, err
	}

	tlsClientConfig := TLSClientConfig{}

	if _, err := certutil.NewPool(rootCAFile); err != nil {
		klog.Errorf("Expected to load root CA config from %s, but got err: %v", rootCAFile, err)
	} else {
		tlsClientConfig.CAFile = rootCAFile
	}

	return &Config{
		// TODO: switch to using cluster DNS.
		Host:            "https://" + net.JoinHostPort(host, port),
		TLSClientConfig: tlsClientConfig,
		BearerToken:     string(token),
		BearerTokenFile: tokenFile,
	}, nil
}
```

There may be a better long term solution but for now I propose we do the following:

* Deploy the operator in the data plane clusters
  * Do not deploy the operator in the control plane cluster yet
* On the client side get the following from each remote cluster
  * service account token (from operator pod service account)
  * CA cert
  * api server url and port which can be obtained from the `KUBERNETES_SERVICE_HOST`, `KUBERNETES_SERVICE_PORT` env vars
* Store the remote cluster credentials in a secret
* Create a ClientConfig that references the secret
  * Add a field in the ClientConfig CRD to indicate whether it is for a kubeconfig or a service account token
* Create a script for extracts credentials and creating the secret and ClientConfig
* The operator can (if necessary) duplicate the logic from the `InClusterConfig()` function in order to create a `rest.Config`

We should do a quick prototype on this to confirm it works.

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-797) by [Unito](https://www.unito.io)
┆Epic: Multi-Cluster Deployment
┆Fix Versions: k8ssandra-2.0.0-alpha
┆Issue Number: K8SSAND-797
┆Priority: Medium

**Comment by jsanda** [comment]
@adutra pointed out this [article](https://goteleport.com/blog/kubectl-gke/) which describes an approach for accessing GKE without requiring `gcloud`. It uses this [script](https://github.com/gravitational/teleport/blob/master/examples/k8s-auth/get-kubeconfig.sh). 

The script is similar to what I described in that it gets credentials from a service account. It differs though in that it creates a kubeconfig entry with those credentials. I like the idea of creating the kubeconfig entry because it involves less changes in the operator and won't require any additional effort by the user.

**Comment by jsanda** [comment]
Closed by #89
